proto/kuksa/val/v2/types.proto

```diff
@@ -0,0 +1,274 @@
+/********************************************************************************
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License 2.0 which is available at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+syntax = "proto3";
+
+package kuksa.val.v2;
+import "google/protobuf/timestamp.proto";
+
+option go_package = "kuksa/val/v2";
+
+// Define which data we want
+message EntryRequest {
+  string path = 1;
+  EntryType entry_type = 2;
+}
+
+// Define the data we want to set
+message EntryUpdate {
+  DataEntry entry = 1;
+}
+
+message DataEntry {
+  // Defines the full VSS path of the entry.
+  string path = 1;
+
+  // The value (datapoint)
+  Datapoint value = 2;
+
+  EntryType entry_type = 3;
+}
+
+message Datapoint {
```

Comment:
Added the ``  oneof state {
    Failure failure,
    Value value
  }`` structure to it so we can add this information.

Reply:
I think this way of specifying missing values is located in the wrong place and / or has the incorrect enumeration of possible errors.

A data point only exists when there is a timestamp associated with it. I.e.

| Timestamp | Value         |
|-----------|---------------|
|    t0     | NOT_AVAILABLE |
|    t1     | 4.32          |
|    t2     | 12.41         |
|    t3     | NOT_AVAILABLE |

If a provider sets the value to `NOT AVAILABLE`, that change has a specific timestamp just the same as if the value was set to `12.41`. Furthermore, this is the timestamp + value that every authorized user would see when looking at that particular signal.

The same can not be said of ACCESS DENIED, as it doesn't have a specific timestamp and would differ between users. Furthermore, different reasons listed in `FailureValue` for a value being `NOT_AVAILABLE` does not change the fact that the value is `NOT_AVAILABLE`.

If these reasons are to be communicated, it would make more sense to do it at another point in the API, i.e. through an `Error` or `Status` one level above the data point as it should not exist without a timestamp.

i.e. something like:
```protobuf
message Response {
  path = 1;
  oneof data {
    DataPoint data_point = 2;
    Error error = 3;
  }
}

message DataPoint {
  google.protobuf.Timestamp timestamp = 1;
  Value value = 2;
}
```

Reply:
You're right regarding the state of a value as it is stored in the broker's db: This could only have three basic states:
* Some valid value
* The value is invalid
* The value is not provided

The other reasons are "just" states as seen from the consumer's perspective.

In your proposal - how to express that the DataPoint doesn't represent a valid value, means that is currently not provided or is invalid?

Reply:
I think this would end up in something like
```
message Response {
  oneof result {
    DataPoint data_point = 1;
    Error error = 2;
  }
}

message DataPoint {
  google.protobuf.Timestamp timestamp = 1;
  oneof value_state {
    Value value = 2;
    ValueFailure = 3;
  }
}

enum Error {
  UNSPECIFIED = 0;
  UNKNOWN_SIGNAL = 1;
  ACCESS_DENIED = 2;
  INTERNAL_ERROR = 3;
}

enum ValueFailure {
  UNSPECIFIED = 0;
  INVALID_VALUE = 1;
  NOT_PROVIDED = 2;
}
```
The thing with this is that it makes things more complicate as the receiver needs to check the two nested oneofs.
Another option to keep things more simple, would be to define the timestamp as optional (to be unset in case of "access denied" or "unknown signal". The other option is to set the timestamp to the point of time when the request was processed.

Reply:
> @argerus: You're right regarding the state of a value as it is stored in the broker's db: This could only have three basic states:
> 
>     * Some valid value
> 
>     * The value is invalid
> 
>     * The value is not provided

I don't agree with this. A data point either:

* has a value.
* does not have a value.

Everything else are just details around the reason for it not having a value.

> I think John's proposal would end up in something like

No. My proposal is as stated above.

> The thing with this is that it makes things more complicate as the receiver needs to check the two nested oneofs.

No. There is either a data point or not. And the data point (if it exists) can either have a value or not. The only difference is that the cases where it doesn't make sense to have a data point at all, will be expressed by having an Error instead of a data point.

If we in addition to this want to express _why_ a data point doesn't have a value, that is something we can add as some form of status enum. But it doesn't change the fundamental functionality.

Reply:
> I don't agree with this. A data point either:
> * has a value.
> * does not have a value.
>
> Everything else are just details describing the reason for it not having a value. 

If one just wants to check the simple fact, if a data point has a valid value or not can be done via calling .hasValue() at the oneof. If one is interested in the details, they can dig deeper into ValueFailure.

>  It is not possible to set an invalid value in databroker.

Then it should be made possible. It can happen, that a signal will have a invalid value (e.g. because of some sensor misbehaviour).

> No. My proposal is as stated above.

That's the point. Imho, this proposal isn't sufficient. How to express the non-availability together with a timestamp? Answering the question "since when is it unavailable"?

> No. There is either a data point or not. And the data point (if it exists) can either have a value or not. The only difference is that the cases where it doesn't make sense to have a data point will be expressed by having an Error instead of a data point.

You describe two levels here, why not getting a value:
1. The signal is not known at all 
2. The signal is known, but does not have a valid value.

> If we in addition to this want to express why a data point doesn't have a value, there are ways to do this that doesn't introduce multiple levels of oneofs. Not sure it would be a particularly useful thing to have though...

Thats, what I wanted to avoid with the initial proposal: Having just one oneof - either having a value or not - plus describing the reason why.

Reply:
> That's the point. Imho, this proposal isn't sufficient. How to express the non-availability together with a timestamp? Answering the question "since when is it unavailable"?

This is perfectly possible in my proposal.

```
DataPoint {
  timestamp: 2024-06-06T07:45:40.556631Z,
  value: None,
}
```

which describes the time at which the value was set to `None`.

Reply:
> which describes the time at which the value was set to None.

How to set a value to `None`? The it should be declared `optional` ...

Reply:
The main point of my change proposal is that it doesn't make sense to have a data point if it doesn't correspond to a specific point in time. Then it is not a data point. That's the reason for _moving_ the `oneof` one level up, so that we can express an error when there is no data point.

It does seem like we both agree that a signal changing it's value to "not available" - for any reason - represents a specific point in time, i.e. a data point.

I'm not proposing having multiple levels of `oneof`s. That concept was introduced by you when you said that my proposal "would end up" that way. I don't see why it would.

Reply:
> So you think it should be possible to set values that are:
>
> *    outside of min / max?
> *    not in the allowed set?
> *    have the wrong data type?

I think, you got me wrong. I explicitly don't want that. I just want to have a state to say that a signal currently does not have a valid value, i.e. a state expressible by the data type plus its defined range. For exa, if you have a speed signal with data type int and unit km/h. How to say, that this signal is currently not giving a valid speed value?

Reply:
> I think, you got me wrong. I explicitly don't want that.

Ah ok. Good :smile: 

So then you _do_ agree that a signal having a value or not is a **binary** thing, i.e. it either has a (valid) value or it doesn't.

But you want a way to express the _reason_ for it not having a value.

Reply:
> How to set a value to None? Then it should be declared optional ...

Not in proto3 (which we are using). All (complex type) fields are optional in proto3.

Reply:
What would the issue be if we had a status field expressing these reasons?
If a consumer doesn't care about _why_ a value is not available, then they don't need to check this at all.

That could also be used in order to subscribe to the status of a signal, without subscribing to the actual value (think high frequency updates, but the consumer only cares whether the signal is alive or not).

```proto

enum SensorStatus {
  STATUS_UNSPECIFIED = 0;
  STATUS_OK = 1;
  STATUS_SOME_ERROR = 2;
  ...
}
```

Reply:
> [...] it doesn't make sense to have a data point if it doesn't correspond to a specific point in time. Then it is not a data point.

Agreed.

> It does seem like we both agree that a signal changing it's value to "not available" - for any reason - represents a specific point in time, i.e. a data point.

Yes.

> I'm not proposing having multiple levels of oneofs. That concept was introduced by you when you said that my proposal "would end up" that way. I don't see why it would.

Because I thought the value could not be `None`. But that seems possible.

> So then you do agree that a signal having a value or not is a binary thing, i.e. it either has a (valid) value or it doesn't.
> But you want a way to express the reason for it not having a value.

Yes, I could imagine that in some cases apps would like to differentiate between the two situations "the value is currently not provided" and "it is provided, but currently not valid". And yes, in many cases this won't be relevant and apps could handle both situations in the same way.

Reply:
Also, imho the situation, that an app does not have the access right to a certain signal, could be also of temporary nature, e.g. if the user temporary disallows an app to access the current vehicle position.

Reply:
> What would the issue be if we had a status field expressing these reasons?

Basically, that's what we have with the oneof.

> If a consumer doesn't care about why a value is not available, then they don't need to check this at all.

With the current proposal it's easy as that. E.g. for C++:
```C++
void handleGetRepsonse(const GetResponse& res) {
    const auto& datapoint = res.getEntry();
    auto timestamp = datapoint.getTimestamp();
    if (datapoint.hasValue()) {
        auto value = datapoint().getValue().getInt32(); // If I know it's an int32
    } else {
        // has no value - that's it if I'm not interested why. Else:
        auto failureReason = datapoint.getFailure();
        switch (failureReason)
        { 
            ...
        }
    }
}
```
> That could also be used in order to subscribe to the status of a signal, without subscribing to the actual value (think high frequency updates, but the consumer only cares whether the signal is alive or not).

I'm not convinced about the need for that ...

Reply:
So you are arguing in favor of having an additional `oneof` used to make `Failure` and `Value` mutually exclusive. I.e. `Failure` would only exist when there is no value.

The `Status` I'm arguing for would instead always exist and would contain `STATUS_OK` unless the value is missing for some reason.

>> That could also be used in order to subscribe to the status of a signal, without subscribing to the actual value (think high frequency updates, but the consumer only cares whether the signal is alive or not).
>
> I'm not convinced about the need for that ...

Ok. But being able to subscribe to - and get a notification - whenever a signal of interest stops having the OK status seems very useful _to me_ at least. Imagine writing a monitoring/logging applications that should trigger a notification whenever a signal goes missing for whatever reason. Subscribing to all signals _values_ in car order to figure out that seems extremely wasteful in terms of bandwidth etc. 

Having a `Status` field is not strictly necessary in order to have such a subscription feature. But it _would_ add consistency, if the `Status` you subscribed to was the same as the `Status` you received doing a `Get`.

I also find it weird that you are not convinced of the need to monitor the status of a signal, while at the same time arguing strongly for the need of differentiating between a signal value being "NOT AVAILABLE" or having "INVALID VALUE". That in my mind is one level less important than finding out that the value is missing in the first place.

Reply:
> So you are arguing in favor of having an additional oneof used to make Failure and Value mutually exclusive. I.e. Failure would only exist when there is no value.

Yes, exactly - but I wouldn't say it's an _additional_ oneof - it's just one. I see this "mutually exclusiveness" as an (small) advantage over something like:
```
message DataPoint {
  Timestamp ts;
  Status status;
  optional Value value;
}
```
With this you could have the situation, that the value is set although the status is not "Ok". Code-wise I don't see a big diff here, because with both solutions the consumer has to check if the value is set or not - and then possibly checking the status for why it doesn't. Small further diff: In state "Ok" you'd transmit both information: the status and the value. In the "oneof" solution always only one thing is transmitted.

But if everyone prefers this over the oneof, it would be also acceptable for me.

> I also find it odd that you are not convinced of the need to subscribe to the status of a signal, while at the same time arguing strongly for the need to be able to differentiate between a signal being "NOT AVAILABLE" or having "INVALID VALUE". That is something I have a harder time seeing the need for.

Saying "I'm not convinced" doesn't mean "I deny it". I was more thinking about a "normal" app working with some set of signals. For those use cases you described, this would make sense, of course. 

Would you then like to separate both? Something like this:
```
rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse);
rpc SubscribeSignalStatus(SubscribeSignalStatusRequest) returns (stream SubscribeSignalStatusResponse);

message SubscribeRequest {
  string path = 1;
}
message SubscribeResponse {
  Datapoint datapoint = 1;
}
message DataPoint {
  google.protobuf.Timestamp timestamp = 1;
  Value value = 2;
}

message SubscribeSignalStatusRequest {
  repeated string paths = 1;
}
message SubscribeSignalStatusResponse {
  map<string, SignalStatus> signal_status = 1;
}
enum SignalStatus {
  UNSPECIFIED = 0;
  PROVIDED = 1;
  NOT_PROVIDED = 2;
  ACCESS_DENIED = 3;
  UNKNOWN_SIGNAL = 4;
  INTERNAL_ERROR = 5;
}
```
(For the "pure" signal status I wouldn't then provide the state representing an "invalid value", because that's something what also could happen quite frequently.)

I'm not against having this "signal status subscribe". But I still would like to have the status also in the "normal" subscribe: I see situations, where the consumer not only needs to know if a value is present or not, but also needs to distinguish why it does not. If I separate those in two subscribes I put the burden of syncing between both to the consumer.